include/alp/imf.hpp

```diff
@@ -50,107 +50,202 @@
+					std::cout << "Select move constructor" << std::endl;
+					std::cout << (this->select).size() << std::endl;
+					//if ( *std::max_element( select.cbegin(), select.cend() ) >= N) {
+					//	throw std::runtime_error("IMF Select beyond range.");
+					//}
+				}
+
+				template< typename OtherIMF >
+				bool isSame( const OtherIMF &other ) const {
+					return IMF::isSame( other ) && select == static_cast< const Select & >( other ).select;
+				}
+		};
+
+		/**
+		 * A composition of two IMFs.
+		 * \f$I_{g,n} =[0, n), I_{g,N} =[0, N)\f$
+		 * \f$I_{f,n} =[0, n), I_{f,N} =[0, N)\f$
+		 * \f$Composed_{f, g} = I_{g,n} \rightarrow I_{f,N}; i \mapsto f( g( i ) )\f$
+		 *
+		 * \tparam FirstIMF  The function that is applied first (i.e., \f$g\f$ )
+		 * \tparam SecondIMF The function that is applied second (i.e., \f$f\f$ )
+		 *
+		 * For specific combinations of the IMF types, there are specializations
+		 * that avoid nested function calls by fusing two functions into one.
+		 */
+
+		template< typename FirstIMF, typename SecondIMF >
+		class Composed: public IMF {
+
+			public:
+				const FirstIMF &g;
+				const SecondIMF &f;
+
+				size_t map( const size_t i ) const {
+						std::cout << "Calling Composed< IMF, IMF>::map()" << std::endl;
+						return f.map( g.map( i ) );
+				}
+
+				Composed( const FirstIMF &g, const SecondIMF &f ):
+					IMF( g.n, f.N ), g( g ), f( f ) {}
+
+		};
+
+		template< typename IMF1, typename IMF2 >
+		struct composed_type {
+			typedef Composed< IMF1, IMF2 > type;
+		};
+
+		template<>
+		struct composed_type< Strided, Strided > {
+			typedef Strided type;
+		};
+
+		template<>
+		struct composed_type< Id, Strided > {
+			typedef Strided type;
+		};
+
+		template<>
+		struct composed_type< Strided, Id > {
+			typedef Strided type;
+		};
+
+		template<>
+		struct composed_type< Id, Id > {
+			typedef Id type;
+		};
+
+		/**
+		 * Creates the composed IMF from two provided input IMFs.
+		 * Depending on the input IMF types, the factory may
+		 * specialize the returned IMF type.
+		 */
+
+		struct ComposedFactory {
+
+			template< typename IMF1, typename IMF2 >
+			static typename composed_type< IMF1, IMF2 >::type create( const IMF1 &f1, const IMF2 &f2 );
+
+		};
+
+		template<>
+		Strided ComposedFactory::create< Id, Strided >( const Id &f1, const Strided &f2 ) {
+			return Strided( f1.n, f2.N, f1.s * f2.s, f1.s * f2.b + f1.b );
+		}
+
+		template<>
+		Strided ComposedFactory::create( const Strided &f1, const Strided &f2 ) {
+			return Strided( f1.n, f2.N, f1.s * f2.s, f1.s * f2.b + f1.b );
+		}
+
+		template<>
+		Strided ComposedFactory::create( const Strided &f1, const Id &f2 ) {
+			return Strided( f1.n, f2.N, f1.s * f2.s, f1.s * f2.b + f1.b );
+		}
+
+		/** Composition of two Id IMFs is an Id Imf */
+		template<>
+		Id ComposedFactory::create( const Id &f1, const Id &f2 ) {
+			(void)f2;
+			// The first function's domain must be equal to the second function's domain.
+			assert( f1.N == f2.n );
```

Comment:
why is one N capital and the other one is a small n?

Reply:
N is the cardinality of the co-domain, n is the cardinality of the domain of a mapping function. For the composition, f2(f1(x)), we require that f1.N == f2.n.
I see that the comment in the code is incorrect wrt the domain/co-domain nomenclature. Will fix

Reply:
ah got it! Thanks

Reply:
the (void)f2; I see now btw could go into an `ifdef NDEBUG`, though that's being very pedantic... I think (but am not sure) that some linting tools will otherwise complain about it (when NDEBUG is not defined)